training3.py

```diff
@@ -0,0 +1,235 @@
+        return score
+
+    # Helper function for get_heuristic: checks if window satisfies heuristic conditions
+    def check_window(window, num_discs, piece, config):
+        return (window.count(piece) == num_discs and window.count(0) == config.inarow - num_discs)
+
+    # Helper function for get_heuristic: counts number of windows satisfying specified heuristic conditions
+    def count_windows(grid, num_discs, piece, config):
+        num_windows = 0
+        # horizontal
+        for row in range(config.rows):
+            for col in range(config.columns - (config.inarow - 1)):
+                window = list(grid[row, col:col + config.inarow])
+                if check_window(window, num_discs, piece, config):
+                    num_windows += 1
+        # vertical
+        for row in range(config.rows - (config.inarow - 1)):
+            for col in range(config.columns):
+                window = list(grid[row:row + config.inarow, col])
+                if check_window(window, num_discs, piece, config):
+                    num_windows += 1
+        # positive diagonal
+        for row in range(config.rows - (config.inarow - 1)):
+            for col in range(config.columns - (config.inarow - 1)):
+                window = list(grid[range(row, row + config.inarow), range(col, col + config.inarow)])
+                if check_window(window, num_discs, piece, config):
+                    num_windows += 1
+        # negative diagonal
+        for row in range(config.inarow - 1, config.rows):
+            for col in range(config.columns - (config.inarow - 1)):
+                window = list(grid[range(row, row - config.inarow, -1), range(col, col + config.inarow)])
+                if check_window(window, num_discs, piece, config):
+                    num_windows += 1
+        return num_windows
+
+    #########################
+    # Agent makes selection #
+    #########################
+
+    # Get list of valid moves
+    valid_moves = [c for c in range(config.columns) if obs.board[c] == 0]
+    # Convert the board to a 2D grid
+    grid = np.asarray(obs.board).reshape(config.rows, config.columns)
+    # Use the heuristic to assign a score to each possible board in the next turn
+    scores = dict(zip(valid_moves, [score_move(grid, col, obs.mark, config) for col in valid_moves]))
+    # Get a list of columns (moves) that maximize the heuristic
+    max_cols = [key for key in scores.keys() if scores[key] == max(scores.values())]
+    # Select at random from the maximize
+
+# env = ConnectFourGym([minimax_agent, 'random'])
+env = ConnectFourGym([minimax_agent])
+env
+
+# # Create directory for logging training information
+# log_dir = "log/"
+# os.makedirs(log_dir, exist_ok=True)
+#
+# # Logging progress
+# env = Monitor(env, log_dir, allow_early_resets=True)
+# env
+
+vec_env = DummyVecEnv([lambda: env])
+vec_env
+
+
+class Net(BaseFeaturesExtractor):
```

Comment:
I think the constructor is wrong. See https://stable-baselines3.readthedocs.io/en/master/guide/custom_policy.html#custom-feature-extractor

Reply:
Should I change features_dim to 256 as it is shown in https://stable-baselines3.readthedocs.io/en/master/guide/custom_policy.html#custom-feature-extractor?

Reply:
This is ok